File: server/src/main/java/pl/ordovita/workspace/domain/port/WorkspaceRepository.java
package pl.ordovita.workspace.domain.port;

import pl.ordovita.identity.domain.model.user.UserId;
import pl.ordovita.workspace.domain.model.Workspace;
import pl.ordovita.workspace.domain.model.WorkspaceId;

import java.util.List;
import java.util.Optional;
import java.util.Set;

public interface WorkspaceRepository {
    Workspace save(Workspace workspace);
    Optional<Workspace> findById(WorkspaceId id);
    Set<Workspace> findAllByCreatedBy(UserId id);
    void delete(Workspace workspace);
}

File: DotNetServer/Ordovita.Infrastructure/Workspace/Persistance/WorkspaceRepository.cs
using Microsoft.EntityFrameworkCore;
using Ordovita.Domain.Identity;
using Ordovita.Domain.Workspace;
using Ordovita.Domain.Workspace.port;
using Ordovita.Infrastructure.Persistence;
using WorkspaceAggregate = Ordovita.Domain.Workspace.Workspace;

namespace Ordovita.Infrastructure.Workspace.Persistance;

public sealed class WorkspaceRepository(AppDbContext context) : IWorkspaceRepository
{
    public async Task AddAsync(WorkspaceAggregate workspace, CancellationToken ct = default)
    {
        await context.Workspaces.AddAsync(workspace, ct);
    }

    public async Task<WorkspaceAggregate?> GetByIdAsync(WorkspaceId id, CancellationToken ct = default)
    {
        return await context.Workspaces
            .Include(w => w.AssignedUsers)
            .FirstOrDefaultAsync(w => w.Id == id, ct);
    }

    public async Task<IReadOnlyList<WorkspaceAggregate>> GetByCreatedByAsync(UserId createdBy, CancellationToken ct = default)
    {
        return await context.Workspaces
            .AsNoTracking()
            .Include(w => w.AssignedUsers)
            .Where(w => w.CreatedBy == createdBy)
            .ToListAsync(ct);
    }

    public void Delete(WorkspaceAggregate workspace)
    {
        context.Workspaces.Remove(workspace);
    }
}
